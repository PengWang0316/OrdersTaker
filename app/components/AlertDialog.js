import React from 'react';
import PropTypes from 'prop-types';
import { Dialog, DialogTitle, DialogContent, DialogContentText, DialogActions, Button } from '@material-ui/core';

export const AlertDialog = ({
  open, onClose, title, content, onSecondButton, onFirstButton, firstButtonText, secondButtonText
}) => (
  <Dialog
    open={open}
    onClose={onClose}
    aria-labelledby="alert-dialog-title"
    aria-describedby="alert-dialog-description"
  >
    <DialogTitle id="alert-dialog-title">{title}</DialogTitle>
    <DialogContent>
      <DialogContentText id="alert-dialog-description">{content}</DialogContentText>
    </DialogContent>
    <DialogActions>
      <Button onClick={onFirstButton} color="primary">
        {firstButtonText}
      </Button>
      <Button onClick={onSecondButton} color="primary" autoFocus>
        {secondButtonText}
      </Button>
    </DialogActions>
  </Dialog>
);
AlertDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  onFirstButton: PropTypes.func.isRequired,
  onSecondButton: PropTypes.func.isRequired,
  firstButtonText: PropTypes.string,
  secondButtonText: PropTypes.string
};
AlertDialog.defaultProps = {
  firstButtonText: 'Cancel',
  secondButtonText: 'Confirm'
};
export default AlertDialog;
