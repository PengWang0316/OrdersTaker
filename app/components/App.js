
import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
// import Loadable from 'react-loadable'; Does not work with react-hot-loader very well.
// import universal from 'react-universal-component'; Does not work with react-hot-loader very well.
import importedComponent from 'react-imported-component';
import CssBaseline from '@material-ui/core/CssBaseline';

import LoginDialog from './LoginDialog/';
import LoginDialogSnackbar from './snackbars/LoginDialogSnackbar';
import LogoutSnackbar from './snackbars/LogoutSnackbar';
import Navbar from './Navbar';
import { parserUserFromJwt } from '../actions/UserActions';
import LoginDialogContext from '../contexts/LoginDialogContext';

import { JWT_MESSAGE, HOME_PAGE_URL, LOGIN_REDIRECT_RUL, ORDER_PAGE_URL, ORDER_STATUS_PAGE_URL, ORDERS_PAGE_URL } from '../config';
import LoadingAnimation from './SharedComponents/LoadingAnimation';

// import { CSSTransitionGroup } from 'react-transition-group';

// import BackgroundMainPager from './BackgroundMainPager';

const theme = createMuiTheme({
  palette: {
    primary: {
      light: '#9ecde7',
      main: '#6d9cb5',
      dark: '#3e6e85',
      contrastText: '#fff',
    },
    secondary: {
      light: '#497ca4',
      main: '#105075',
      dark: '#002884',
      contrastText: '#fff',
    },
  },
});

/* istanbul ignore next */
const HomePage = importedComponent(() => import(/* webpackChunkName: "HomePageContainer" */ './containers/HomePageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const LoginRedirect = importedComponent(() => import(/* webpackChunkName: "LoginRedirect" */ './LoginRedirect').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrderPage = importedComponent(() => import(/* webpackChunkName: "OrderPageContainer" */ './containers/OrderPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrderStatusPage = importedComponent(() => import(/* webpackChunkName: "OrderStatusPageContainer" */ './containers/OrderStatusPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrdersPage = importedComponent(() => import(/* webpackChunkName: "OrdersPageContainer" */ './containers/OrdersPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/**
 * The root component that contains the theme, routers, navbar, and login dialog
 */
export class App extends Component {
  static propTypes = {
    user: PropTypes.object.isRequired,
    parserUserFromJwt: PropTypes.func.isRequired
  };

  /**
   * If Redux does not have user and local storage has jwt, use jwt to parser the user.
   * @param {object} props contains the component's prop value.
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    if (!props.user._id) {
      const jwtMessage = localStorage.getItem(JWT_MESSAGE);
      if (jwtMessage) props.parserUserFromJwt(jwtMessage);
    }
  }

  state = {
    isLoginDialogOpen: false,
    isLogoutSnackBarOpen: false,
    isLoginSnackbarOpen: false,
    snackbarMessage: ''
  };

  /**
   * Setting the open state to an opposite value.
   * @return {null} No return.
   */
  handleToggleDialog = () => this.setState(({ isLoginDialogOpen }) => ({ isLoginDialogOpen: !isLoginDialogOpen }));

  /**
   * Setting the snackbarOpen state to true in order to open the snackbar.
   * If message paramter is a string, set it to message state. Otherwise, set a empty message.
   * @param {string} message will be showed in the snackbar.
   * @return {null} No return.
   */
  handleToggleSnackbar = message => this.setState(({ isLoginSnackbarOpen }) =>
    ({ isLoginSnackbarOpen: !isLoginSnackbarOpen, snackbarMessage: typeof message === 'string' ? message : '' }));

  /**
   * Set the logoutSnackbarOpen state to an opposite value.
   * @return {null} No return.
   */
  handleToggleLogoutSnackbar = () => this.setState(({ isLogoutSnackBarOpen }) =>
    ({ isLogoutSnackBarOpen: !isLogoutSnackBarOpen }));

  /**
   * It will be pass down in a context in order to make it available for all component.
   * This method will be used when logout action is performed.
   * Clear the state isLoginDialogOpen in order to prevent the LoginDialog shows up.
   * @return {null} No return.
   */
  handleLogoutAction = () => this.setState({ isLoginDialogOpen: false, isLogoutSnackBarOpen: true });

  /**
   * The render method for the component
   * @return {jsx} Return jsx
   */
  render() {
    const { user } = this.props;
    const { isLogoutSnackBarOpen, isLoginDialogOpen, isLoginSnackbarOpen, snackbarMessage } = this.state;
    return (
      <MuiThemeProvider theme={theme}>
        <Router>
          {/* Using a context to give children components access to login dialog */}
          <LoginDialogContext.Provider value={{ handleLogoutAction: this.handleLogoutAction, handleToggleLoginDialog: this.handleToggleDialog }}>
            <div>
              <CssBaseline />
              <Navbar />
              <main>
                <Switch>
                  <Route exact path={HOME_PAGE_URL} component={HomePage} />
                  <Route exact path={ORDER_PAGE_URL} component={OrderPage} />
                  <Route exact path={LOGIN_REDIRECT_RUL} component={LoginRedirect} />
                  <Route path={`${ORDER_STATUS_PAGE_URL}/:orderId`} component={OrderStatusPage} />
                  <Route path={ORDERS_PAGE_URL} component={OrdersPage} />
                  <Route render={() => <p>Not Fount!</p>} />
                </Switch>
              </main>
              {!user._id && ( // If a user has already login, do not need to load componnets below.
                <Fragment>
                  <LoginDialog open={isLoginDialogOpen} onClose={this.handleToggleDialog} onToggleSnackbar={this.handleToggleSnackbar} />
                  <LoginDialogSnackbar open={isLoginSnackbarOpen} onClose={this.handleToggleSnackbar} message={snackbarMessage} />
                </Fragment>
              )}
              <LogoutSnackbar open={isLogoutSnackBarOpen} onClose={this.handleToggleLogoutSnackbar} />
            </div>
          </LoginDialogContext.Provider>
        </Router>
      </MuiThemeProvider>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  user: state.user
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  parserUserFromJwt: jwtMessage => dispatch(parserUserFromJwt(jwtMessage))
});
export default connect(mapStateToProps, mapDispatchToProps)(App);
