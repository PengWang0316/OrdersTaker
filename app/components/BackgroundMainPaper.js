import React from 'react';
import PropTypes from 'prop-types';
import { withStyles, Paper } from '@material-ui/core';

const styles = {
  root: {
    backgroundImage: 'linear-gradient(to right top, #051937, #004d7a, #008793, #00bf72, #a8eb12)',
    width: '100%',
    height: '100%',
    padding: '0 20px'
  },
  paper: {
    margin: '40px auto 20px auto',
    maxWidth: 1024,
    borderRadius: '10px'
  }
};

export const BackgroundMainPaper = ({ classes, children }) => (
  <div className={classes.root}>
    <Paper className={classes.paper} elevation={4}>
      {children}
    </Paper>
  </div>
);
BackgroundMainPaper.propTypes = {
  classes: PropTypes.object.isRequired,
  children: PropTypes.element.isRequired
};
export default withStyles(styles)(BackgroundMainPaper);
