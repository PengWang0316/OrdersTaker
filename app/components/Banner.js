import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Hidden from '@material-ui/core/Hidden';
import { withStyles } from '@material-ui/core/styles';

import { fetchBasicInformation } from '../actions/BasicInformationActions';

const styles = {
  divPlaceHolder: {
    background: '#efefef',
    width: '100%',
    minHeight: '400px'
  },
  bigBanner: {
    width: '100%',
    maxHeight: '770px'
  },
  smallBanner: {
    width: '100%',
    maxHeight: '400px'
  },
  bannerText: {
    display: 'flex',
    flexDirection: 'column',
    width: '60%',
    position: 'absolute',
    marginTop: '18%',
    left: '20%'
  },
  title: {
    fontSize: '40px',
    color: 'white',
    textAlign: 'center',
    textTransform: 'uppercase',
    textShadow: '1px 1px 1px #909090'
  },
  subTitle: {
    color: 'white',
    fontSize: '24px',
    textAlign: 'center',
    textShadow: '1px 1px 1px #909090'
  },
  '@media (max-width: 600px)': {
    bannerText: { marginTop: '21%' },
    title: { fontSize: '30px' },
    subTitle: { fontSize: '18px' },
    divPlaceHolder: { minHeight: '230px' }
  }
};

/**
 * The Component to show the banner image and title.
 */
export class Banner extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    basicInformation: PropTypes.object,
    fetchBasicInformation: PropTypes.func.isRequired
  };
  static defaultProps = { basicInformation: null };

  /**
   * Call the redux action to get the basic infomariont.
   * @param {object} props contains this component's prop values;
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    if (!props.basicInformation) props.fetchBasicInformation();
  }

  /**
   * Render the jsx for the component.
   * @return {jsx} Return the jsx.
   */
  render() {
    const { classes, basicInformation } = this.props;
    return (
      <div className={basicInformation ? '' : classes.divPlaceHolder}>
        {basicInformation &&
          <Fragment>
            <div className={classes.bannerText}>
              <div className={classes.title}>{basicInformation.banners[0].title}</div>
              <div className={classes.subTitle}>{basicInformation.banners[0].subTitle}</div>
            </div>
            <Hidden only="xs">
              <img alt="Big Banner" src={basicInformation.banners[0].url} className={classes.bigBanner} />
            </Hidden>
            <Hidden only={['lg', 'md', 'sm']}>
              <img alt="Small Banner" src={basicInformation.banners[0].xsUrl} className={classes.smallBanner} />
            </Hidden>
          </Fragment>}
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = ({ basicInformation }) => ({ basicInformation });
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchBasicInformation: () => dispatch(fetchBasicInformation())
});
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(Banner));
