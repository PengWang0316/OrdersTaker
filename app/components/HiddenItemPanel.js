import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Button } from '@material-ui/core';
import { ExpandMore, ExpandLess } from '@material-ui/icons';
import { scroller } from 'react-scroll';

import MenuItem from './MenuItem';

const flexBox = {
  flexWrap: 'wrap',
  height: 'auto',
  maxHeight: 0,
  // lineHeight: 0,
  display: 'flex',
  overflow: 'hidden',
  transition: 'max-height 0.7s'
};

const styles = {
  buttonDiv: {
    width: '100%',
    textAlign: 'right'
  },
  flexBoxAround: {
    ...flexBox,
    justifyContent: 'space-around'
  },
  // flexBoxStart: {
  //   ...flexBox,
  //   justifyContent: 'flex-start'
  // },
  '@media (max-width: 500px)': {
    flexBoxAround: { justifyContent: 'center' }
  }
};

/**
 * Show the loading button and extra items
 */
export class HiddenItemPanel extends Component {
  static propTypes = {
    items: PropTypes.array.isRequired,
    category: PropTypes.string.isRequired,
    // itemAmount: PropTypes.number.isRequired,
    classes: PropTypes.object.isRequired
  };

  state = { isExpand: false };

  /**
   * Setting the isExpand state to true and scroll to the button's position when a user click the button.
   * react-scroll library is used.
   * @return {null} No return.
   */
  handleShowMoreClick = () => {
    // this.categoryTitle = this.categoryTitle || document.getElementById(this.props.category);
    this.setState(({ isExpand }) => {
      // if (isExpand) this.categoryTitle.scrollIntoView();
      if (isExpand) scroller.scrollTo(this.props.category, {
        duration: 800,
        delay: 0,
        smooth: 'easeInOutQuart'
      });
      return { isExpand: !isExpand };
    });
  }

  /**
   * Rendering the jsx
   * @return {jsx} Return jsx.
   */
  render() {
    const { items, classes } = this.props;
    const { isExpand } = this.state;
    const divHeight = isExpand ? `${287 * (items.length + 1)}px` : 0; // Max height equal items' number times each element's height(287px).

    return (
      <Fragment>
        {items.length !== 0 && (
          <Fragment>
            <div className={classes.flexBoxAround} style={{ maxHeight: divHeight }}>
              {items.map(item => <MenuItem key={item._id} item={item} />)}
            </div>
            <div className={classes.buttonDiv}>
              <Button color="primary" onClick={this.handleShowMoreClick}>{isExpand ? (<Fragment>Hide <ExpandLess /></Fragment>) : (<Fragment>Show More <ExpandMore /></Fragment>)}</Button>
            </div>
          </Fragment>
        )}
      </Fragment>);
  }
}
export default withStyles(styles)(HiddenItemPanel);
