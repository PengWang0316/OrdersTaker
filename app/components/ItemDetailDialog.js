import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Dialog, Avatar, Typography, DialogContent, Button, IconButton } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { Close } from '@material-ui/icons/';

import PriceItem from './PriceItem';
import RecipePanel from './RecipePanel';
import dialogStyles from '../styles/DialogSytels';

const styles = ({
  ...dialogStyles, // Extending some basic dialog styles from DialogSytels.
  flexBox: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'center'
  },
  avatar: {
    width: '150px',
    height: '150px',
    marginBottom: '8px',
    transition: 'all 0.4s',
    marginRight: 30
  },
  title: {
    fontSize: 18,
    fontWeight: 'bold'
  },
  description: {
    maxWidth: 600,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-around'
  },
  '@media (max-width: 923px)': {
    avatar: {
      width: '200px',
      height: '200px',
      marginBottom: 25,
      marginRight: 0
    }
  }
});

export const ItemDetailDialog = ({
  onClose, open, item, classes
}) => (
  <Dialog onClose={onClose} open={open} classes={{ paper: classes.dialogPaper }}>
    <DialogContent className={classes.dialogContent}>
      {item && (
        <Fragment>
          <IconButton onClick={onClose} className={classes.dialogCloseButton}><Close className={classes.dialogCloseIcon} /></IconButton>
          <div className={classes.flexBox}>
            <Avatar src={item.photo} className={classes.avatar} />
            <div className={classes.description}>
              <Typography color="primary" className={classes.title}>{item.name}</Typography>
              <Typography color="textSecondary">{item.description}</Typography>
              <PriceItem item={item} flexEnd />
            </div>
          </div>
          <RecipePanel recipes={item.recipes} />
          <div className={classes.dialogButtonDiv}>
            <Button onClick={onClose} variant="contained" size="small" aria-label="Close the detail page" color="primary">Close</Button>
          </div>
        </Fragment>
      )}
    </DialogContent>
  </Dialog>
);
ItemDetailDialog.propTypes = {
  onClose: PropTypes.func.isRequired,
  open: PropTypes.bool.isRequired,
  item: PropTypes.object,
  classes: PropTypes.object.isRequired
};
ItemDetailDialog.defaultProps = { item: null };
export default withStyles(styles)(ItemDetailDialog);
