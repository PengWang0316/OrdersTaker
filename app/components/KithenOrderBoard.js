import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  Paper, ExpansionPanel, ExpansionPanelDetails, Typography, ExpansionPanelSummary
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { ExpandMore as ExpandMoreIcon } from '@material-ui/icons';

import { fetchUnfinishedOrders } from '../actions/OrdersActions';
import UnfinishedOrderRow from './UnfinishedOrder/UnfinishedOrderRow';
import UnfinishedOrderList from './UnfinishedOrder/UnfinishedOrderList';

const styles = {
  paper: {
    padding: 10
  }
};

/**
 * The component to show the kithen's order board
 */
export class KithenOrderBoard extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired
  };

  state = {
    unfinishedOrders: null
  };

  /**
   * Fetching the unfinished orders.
   * @param {object} props contains all component's prop value.
   * @return {null} No return.
   */
  componentDidMount() {
    fetchUnfinishedOrders(this.props.user).then(data => this.setState({ unfinishedOrders: KithenOrderBoard.getObjectFromArray(data) }));
  }

  /**
   * Transfer an array to an object with id as the key.
   * @param {array} array is the array will be transferred.
   * @return {object} Return an object and use _id for the key.
   */
  static getObjectFromArray = array => {
    const object = {};
    array.forEach(element => { object[element._id] = element; });
    return object;
  };

  /**
   * Updating the database for the item that is finished and update the unfinishedOrders state.
   * @param {string} orderId is the id of the order is changing.
   * @param {string} itemId is the id of the item is changing.
   * @return {null} No return.
   */
  handleItemClick = (orderId, itemId) => {
    console.log(orderId, itemId);

  };

  /**
   * The render method.
   * @return {jsx} Return the jsx for the component
   */
  render() {
    const { classes } = this.props;
    const { unfinishedOrders } = this.state;
    return (
      <Paper className={classes.paper}>
        {unfinishedOrders && Object.keys(unfinishedOrders).map(key => (
          <ExpansionPanel key={key}>
            <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
              <UnfinishedOrderRow order={unfinishedOrders[key]} />
            </ExpansionPanelSummary>
            <ExpansionPanelDetails>
              <UnfinishedOrderList order={unfinishedOrders[key]} onClick={this.handleItemClick} />
            </ExpansionPanelDetails>
          </ExpansionPanel>
        ))}
      </Paper>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  user: state.user
});
/* istanbul ignore next */
// const mapDispatchToProps = dispatch => ({
// });
export default connect(mapStateToProps, null)(withStyles(styles)(KithenOrderBoard));
