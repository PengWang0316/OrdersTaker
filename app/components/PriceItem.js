import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Typography, IconButton, Tooltip, Menu, MenuItem } from '@material-ui/core';
import { AddBox, KeyboardArrowDown } from '@material-ui/icons';

const priceDiv = {
  display: 'flex',
  justifyContent: 'space-evenly',
  alignItems: 'center'
};

const styles = {
  addButton: {
    width: 40,
    height: 40
  },
  icon: {
    width: 20,
    height: 20
  },
  upperCase: {
    textTransform: 'uppercase'
  },
  priceDiv,
  multiplePriceDiv: {
    ...priceDiv,
    '&:hover': {
      cursor: 'pointer'
    }
  }
};

/**
 * Show the single price or a multiple price menu.
 */
export class PriceItem extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    item: PropTypes.object.isRequired
  };

  state = {
    anchorEl: null
  }

  /**
   * Changing anchorEl state to an click target element or null.
   * @return {null} No Return.
   */
  handleMenuIconClick = ({ currentTarget }) =>
    this.setState(({ anchorEl }) => {
      if (anchorEl) return { anchorEl: null };
      return { anchorEl: currentTarget };
    });

  /**
   * Rendering the jsx for the component.
   * @return {jsx} Return jsx.
   */
  render() {
    const { classes, item } = this.props;
    const { anchorEl } = this.state;
    const pricesObjectKeys = Object.keys(item.prices); // Save the keys of prices.
    if (pricesObjectKeys.length === 1) return (
      <div className={classes.priceDiv}>
        <Typography color="textSecondary">$ {item.prices[pricesObjectKeys[0]]}</Typography>
        <Tooltip id="tooltip-fab" title="Add to your order" placement="right-end">
          <IconButton className={classes.addButton} aria-label="Add to your order" color="primary">
            <AddBox className={classes.icon} />
          </IconButton>
        </Tooltip>
      </div>
    );
    // Show a drop down menu and all price with add buttons.
    return (
      <Fragment>
        <Tooltip id="tooltip-fab" title="Check prices" placement="right-end">
          <div className={classes.multiplePriceDiv} onClick={this.handleMenuIconClick} role="button" tabIndex="-1">
            <Typography color="textSecondary">More Choices</Typography>
            <KeyboardArrowDown className={classes.icon} color="primary" />
          </div>
        </Tooltip>
        <Menu
          id="simple-menu"
          anchorEl={anchorEl}
          open={Boolean(anchorEl)}
          onClose={this.handleMenuIconClick}
        >
          {pricesObjectKeys.map(key => (
            <MenuItem onClick={this.handleMenuIconClick} className={classes.priceDiv} key={key}>
              <Typography color="primary" className={classes.upperCase}>{key}&nbsp;&nbsp;</Typography>
              <Typography color="primary">${item.prices[key]}</Typography>
              <IconButton className={classes.addButton} aria-label="Add to your order" color="primary">
                <AddBox className={classes.icon} />
              </IconButton>
            </MenuItem>
          ))}
        </Menu>
      </Fragment>
    );
  }
}
export default withStyles(styles)(PriceItem);
