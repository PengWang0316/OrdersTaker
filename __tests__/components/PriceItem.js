import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { PriceItem } from '../../app/components/PriceItem';

jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/core/Tooltip', () => 'Tooltip');
jest.mock('@material-ui/core/Menu', () => 'Menu');
jest.mock('@material-ui/core/MenuItem', () => 'MenuItem');

jest.mock('@material-ui/icons/AddBox', () => 'AddBox');
jest.mock('@material-ui/icons/KeyboardArrowDown', () => 'KeyboardArrowDown');

jest.mock('../../app/utils/AnimationUtil', () => ({ animateOrderNumber: jest.fn() }));

describe('PriceItem', () => {
  const defaultProps = {
    item: { prices: { small: 100 } },
    classes: {
      addButton: 'addButton',
      icon: 'icon',
      upperCase: 'upperCase',
      priceDiv: 'priceDiv',
      multiplePriceDiv: 'multiplePriceDiv',
      flexEnd: 'flex-end'
    },
    addItemToCart: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<PriceItem {...props} />);

  test('Initial the component', () => {
    const component = getShallowComponent();
    expect(component.state('anchorEl')).toBeNull();
  });

  test('handleMenuIconClick', () => {
    const component = getShallowComponent();
    expect(component.state('anchorEl')).toBeNull();
    component.instance().handleMenuIconClick({ currentTarget: 'currentTarget' });
    expect(component.state('anchorEl')).toEqual('currentTarget');
    component.instance().handleMenuIconClick({ currentTarget: 'currentTarget' });
    expect(component.state('anchorEl')).toBeNull();
  });

  test('handleAddToCartClick', () => {
    const component = getShallowComponent();
    const AnimationUtil = require('../../app/utils/AnimationUtil');
    component.instance().handleAddToCartClick('priceKey');
    expect(defaultProps.addItemToCart).toHaveBeenCalledTimes(1);
    expect(defaultProps.addItemToCart).toHaveBeenLastCalledWith({ priceKey: 'priceKey', item: defaultProps.item });
    expect(AnimationUtil.animateOrderNumber).toHaveBeenCalledTimes(1);
  });

  test('IconButtons are clicked', () => {
    let component = getShallowComponent();
    const mockClickFn = jest.fn();
    component.instance().handleAddToCartClick = mockClickFn;
    component.find('IconButton').simulate('click');
    expect(mockClickFn).toHaveBeenCalledTimes(1);
    expect(mockClickFn).toHaveBeenLastCalledWith('small');

    component = getShallowComponent({ ...defaultProps, item: { prices: { middle: 100, large: 200 } } });
    component.instance().handleAddToCartClick = mockClickFn;
    component.find('IconButton').at(0).simulate('click');
    expect(mockClickFn).toHaveBeenCalledTimes(2);
    expect(mockClickFn).toHaveBeenLastCalledWith('middle');
    component.find('IconButton').at(1).simulate('click');
    expect(mockClickFn).toHaveBeenCalledTimes(3);
    expect(mockClickFn).toHaveBeenLastCalledWith('large');
  });

  test('Snapshot with one price not flexEnd', () => expect(renderer.create(<PriceItem {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot with one price flexEnd', () => expect(renderer.create(<PriceItem {...{ ...defaultProps, flexEnd: true }} />).toJSON()).toMatchSnapshot());

  test('Snapshot with multiple prices not flexEnd', () => expect(renderer.create(<PriceItem {...{ ...defaultProps, item: { prices: { small: 100, middle: 200, large: 300 } } }} />).toJSON())
    .toMatchSnapshot());
  test('Snapshot with multiple prices flexEnd', () => expect(renderer.create(<PriceItem {...{ ...defaultProps, flexEnd: true, item: { prices: { small: 100, middle: 200, large: 300 } } }} />).toJSON())
    .toMatchSnapshot());
});
