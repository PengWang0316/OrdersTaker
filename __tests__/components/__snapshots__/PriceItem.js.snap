// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceItem Snapshot with multiple prices 1`] = `
Array [
  <Tooltip
    id="tooltip-fab"
    placement="right-end"
    title="Check prices"
  >
    <div
      className="multiplePriceDiv"
      onClick={[Function]}
      role="button"
      tabIndex="-1"
    >
      <Typography
        color="textSecondary"
      >
        More Choices
      </Typography>
      <KeyboardArrowDown
        className="icon"
        color="primary"
      />
    </div>
  </Tooltip>,
  <Menu
    anchorEl={null}
    id="simple-menu"
    onClose={[Function]}
    open={false}
  >
    <MenuItem
      className="priceDiv"
      onClick={[Function]}
    >
      <Typography
        className="upperCase"
        color="primary"
      >
        small
          
      </Typography>
      <Typography
        color="primary"
      >
        $
        100
      </Typography>
      <IconButton
        aria-label="Add to your order"
        className="addButton"
        color="primary"
      >
        <AddBox
          className="icon"
        />
      </IconButton>
    </MenuItem>
    <MenuItem
      className="priceDiv"
      onClick={[Function]}
    >
      <Typography
        className="upperCase"
        color="primary"
      >
        middle
          
      </Typography>
      <Typography
        color="primary"
      >
        $
        200
      </Typography>
      <IconButton
        aria-label="Add to your order"
        className="addButton"
        color="primary"
      >
        <AddBox
          className="icon"
        />
      </IconButton>
    </MenuItem>
    <MenuItem
      className="priceDiv"
      onClick={[Function]}
    >
      <Typography
        className="upperCase"
        color="primary"
      >
        large
          
      </Typography>
      <Typography
        color="primary"
      >
        $
        300
      </Typography>
      <IconButton
        aria-label="Add to your order"
        className="addButton"
        color="primary"
      >
        <AddBox
          className="icon"
        />
      </IconButton>
    </MenuItem>
  </Menu>,
]
`;

exports[`PriceItem Snapshot with one price 1`] = `
<div
  className="priceDiv"
>
  <Typography
    color="textSecondary"
  >
    $ 
    100
  </Typography>
  <Tooltip
    id="tooltip-fab"
    placement="right-end"
    title="Add to your order"
  >
    <IconButton
      aria-label="Add to your order"
      className="addButton"
      color="primary"
    >
      <AddBox
        className="icon"
      />
    </IconButton>
  </Tooltip>
</div>
`;
